content/blog/2022-04-26-Superoptimizer.md

```diff
@@ -0,0 +1,83 @@
++++
+title = "Superoptimizer: a look at the smallest program"
+[[extra.authors]]
+name = "Jonathan Tran"
+[[extra.authors]]
+name = "Victor Giannakouris"
++++
+
+## Introduction
+
+The majority of compilers utilize optimization to improve code, in terms of runtime performance, code size, and other factors. However, optimization does not necessarily generate the optimal program; the optimization process merely generates a better candidate program. In the Superoptimization paper by Massalin,, a different paradigm is introduced: optimization is performed to generate the minimal sized program, producing programs that are smaller compared to traditionally compiler-optimizer programs, as well as human hand-optimized programs [^1].
+
+As an illustration of Superoptimizer’s capabilities, consider an example centered on compiling the signum function, which is defined as follows: 
+
+```int signum(x) {
+	If (x > 0) return 1;
+	Else if (x <  0) return -1;
+	Else return 0;
+}
+```
+
+The author notes this program usually compiles to around 9 instructions, including several conditional jumps, while a human can manually optimize this program to 6 instructions. Remarkably, Superoptimizer finds an even smaller solution, consisting of 4 instructions, which is shown below.
+
+```
+Add.l d0, d0
+Subx.l d1, d1
+Negx.l d0
+Addx.l d1, d1
```

Comment:
TBH it also seems slightly weird to see capitalization in assembly listings, which are almost always all-lowercase.